source/common/config/new_grpc_mux_impl.cc

```diff
@@ -44,12 +59,37 @@ ScopedResume NewGrpcMuxImpl::pause(const std::vector<std::string> type_urls) {
   });
 }
 
+void NewGrpcMuxImpl::registerVersionedTypeUrl(const std::string& type_url) {
+
+  TypeUrlMap& type_url_map = typeUrlMap();
+  if (type_url_map.find(type_url) != type_url_map.end()) {
+    return;
+  }
+  // If type_url is v3, earlier_type_url will contain v2 type url.
+  absl::optional<std::string> earlier_type_url = ApiTypeOracle::getEarlierTypeUrl(type_url);
+  // Register v2 to v3 and v3 to v2 type_url mapping in the hash map.
+  if (earlier_type_url.has_value()) {
+    type_url_map[earlier_type_url.value()] = type_url;
+    type_url_map[type_url] = earlier_type_url.value();
+  }
+}
+
 void NewGrpcMuxImpl::onDiscoveryResponse(
     std::unique_ptr<envoy::service::discovery::v3::DeltaDiscoveryResponse>&& message,
     ControlPlaneStats&) {
   ENVOY_LOG(debug, "Received DeltaDiscoveryResponse for {} at version {}", message->type_url(),
             message->system_version_info());
   auto sub = subscriptions_.find(message->type_url());
+  // If this type url is not watched, try another version type url.
+  if (enable_type_url_downgrade_and_upgrade_ && sub == subscriptions_.end()) {
+    ENVOY_LOG(debug, "fuck.\n");
```

Comment:
nit: not needed, or we can use more clear message.

Reply:
Sorry, forgot to clean up. It is for debug...